### PR TITLE
Added StateMachine and FiniteStateMachine

### PR DESCRIPTION
**Pull request by AndreasHeine**

### CODE REVIEW

File: asyncua/common/statemachine.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        effectivetransitiontime: EffectiveTransitionTime specifies the time when the current state or one of its substates was entered. 
+        If, for example, a StateA is active and – while active – switches several times between its substates SubA and SubB, 
+        then the TransitionTime stays at the point in time where StateA became active whereas the EffectiveTransitionTime changes 
+        with each change of a substate.
+        '''
+        def __init__(self, name, id, number, node=None):
+            self.name = name
+            self.id = id
+            self.number = number
+            self._transitiontime = datetime.datetime.utcnow() #will be overwritten from _write_transition()
```
[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:101]: initialize the statemachine (change to the initstate) is also a transition in my oppinion

File: asyncua/common/statemachine.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+            transition._transitiontime = datetime.datetime.utcnow()
+        transition._effectivetransitiontime = datetime.datetime.utcnow()
+        await self._last_transition_node.write_value(ua.LocalizedText(transition.name, self.locale), ua.VariantType.LocalizedText)
+        if self._optionals:
+            if self._last_transition_id_node:
+                await self._last_transition_id_node.write_value(transition.id, ua.VariantType.String)
+            if self._last_transition_name_node:
+                await self._last_transition_name_node.write_value(ua.QualifiedName(transition.name, self._idx), ua.VariantType.QualifiedName)
+            if self._last_transition_number_node:
+                await self._last_transition_number_node.write_value(transition.id, ua.VariantType.UInt32)
```
[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py]: ty missed that one!

File: asyncua/common/statemachine.py [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+        https://reference.opcfoundation.org/v104/Core/docs/Part5/B.4.3/
+        name: type string will be converted automatically to qualifiedname 
+            -> Name is a QualifiedName which uniquely identifies the current state within the StateMachineType.
+        id: Id is a name which uniquely identifies the current state within the StateMachineType. A subtype may restrict the DataType.
+        number: Number is an integer which uniquely identifies the current state within the StateMachineType.
+        '''
+        def __init__(self, name, id, number, node=None):
+            self.name = name
+            self.id = id
+            self.number = number
```
[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:78]: State and Transition Subclasses are for implementing StateVaraibleType and TransitionVariableType and yes you are right that they are optional, but i would not implement it without due to the fact that the opc ua way for implementing information in the addresspace is my focus and for that a name as browsename of the state node is better then childs with just ids... but i am open for discussions!

File: asyncua/common/statemachine.py [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+            -> Name is a QualifiedName which uniquely identifies a transition within the StateMachineType.
+        id: Id is a name which uniquely identifies a Transition within the StateMachineType. A subtype may restrict the DataType.
+        number: Number is an integer which uniquely identifies the current state within the StateMachineType.
+        transitiontime: TransitionTime specifies when the transition occurred.
+        effectivetransitiontime: EffectiveTransitionTime specifies the time when the current state or one of its substates was entered. 
+        If, for example, a StateA is active and – while active – switches several times between its substates SubA and SubB, 
+        then the TransitionTime stays at the point in time where StateA became active whereas the EffectiveTransitionTime changes 
+        with each change of a substate.
+        '''
+        def __init__(self, name, id, number, node=None):
```
[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:97]: same comment like above!

**Comment by AndreasHeine** [comment]
@swamper123 lets talk tomorow!

**Review by swamper123 — COMMENTED** [review]
Just found that one at the moment. Good job! 👍

File: asyncua/common/statemachine.py [hunk 5]
```diff
@@ -162,12 +168,12 @@ async def init(self, statemachine):
                     _logger.warning(f"{await statemachine.read_browse_name()} LastTransition Unknown propertie: {dn.Text}")
         self._evgen = await self._server.get_event_generator(self.evtype, self._state_machine_node)
 
-    async def change_state(self, state, transition=None, event_msg=None, severity=500):
+    async def change_state(self, state: State, transition: Transition=None, event_msg:str or ua.LocalizedText=None, severity: int=500):
```
[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py:171]: ok good point

[the remark this answers is not in the document]
Comment (swamper123) [reply asyncua/common/statemachine.py]: `from typing import Union`

[the remark this answers is not in the document]
Comment (AndreasHeine) [reply asyncua/common/statemachine.py]: ...

**Comment by oroulet** [comment]
so you want to merge?

**Comment by AndreasHeine** [comment]
from my point maybe tomorrow! 
will improve some examples in the evening... just to show some more variance
@swamper123 are you done with the review?

**Comment by swamper123** [comment]
@AndreasHeine I will take a look on it again today and let you know.

**Review by swamper123 — CHANGES_REQUESTED** [review]
Beside some smaller style issues (there are more, maybe throw a linter on that and let the IDE reformat it), there are a few discussions points which should be solved.

**Comment by AndreasHeine** [comment]
@oroulet ready to merge!

**Pull request merged**